# Part 0. Data Files

## 0.1 Download Data  

In [ ]:
!pip install wget  

In [ ]:
import wget

link_to_data = 'https://github.com/tulip-lab/sit742/raw/master/Assessment/2019/data/wine.json'
DataSet = wget.download(link_to_data)

link_to_data = 'https://github.com/tulip-lab/sit742/raw/master/Assessment/2019/data/stopwords.txt'

DataSet = wget.download(link_to_data)

In [ ]:
!ls

 HighFreq.txt	 statisticByState.csv  'wine (1).json'
 sample_data	'stopwords (1).txt'     wine.csv
 Shirazkey.txt	 stopwords.txt	        wine.json


## 0.2 Load Data  

In [ ]:
import json
import pandas as pd
import matplotlib.pyplot as plt

In [ ]:
file = 'wine.json'

In [ ]:
# write your code here
# read the json file and drop rows with invalid values in the attributes of 'points' and 'price'.

with open('wine.json') as json_p:
    json_dict = json.load(json_p)
    json_dict
    load_Data = pd.DataFrame(json_dict)
    filtered_dat = load_Data[(load_Data['points'].notnull()) & (load_Data['price'].notnull())]
    print(filtered_dat)

            country                                        description  \
1          Portugal  This is ripe and fruity, a wine that is smooth...   
2                US  Tart and snappy, the flavors of lime flesh and...   
3                US  Pineapple rind, lemon pith and orange blossom ...   
4                US  Much like the regular bottling from 2012, this...   
5             Spain  Blackberry and raspberry aromas show a typical...   
6             Italy  Here's a bright, informal red that opens with ...   
7            France  This dry and restrained wine offers spice in p...   
8           Germany  Savory dried thyme notes accent sunnier flavor...   
9            France  This has great depth of flavor with its fresh ...   
10               US  Soft, supple plum envelopes an oaky structure ...   
11           France  This is a dry wine, very spicy, with a tight, ...   
12               US  Slightly reduced, this wine offers a chalky, t...   
14               US  Building on 150 y

# Part 1: numeric anaysis

## 1.1 Explore the data distribution for each column.

In [ ]:
# write your code here
# you may use functions such as describe() on each attribute
filtered_dat.describe(include='all')

country                                        description designation  \
count   120916                                             120975       86196   
unique      42                                             111567       35776   
top         US  Stalky aromas suggest hay and green herbs, wit...     Reserve   
freq     54265                                                  3        1980   
mean       NaN                                                NaN         NaN   
std        NaN                                                NaN         NaN   
min        NaN                                                NaN         NaN   
25%        NaN                                                NaN         NaN   
50%        NaN                                                NaN         NaN   
75%        NaN                                                NaN         NaN   
max        NaN                                                NaN         NaN   

        points          price    province     region_1       region_2  \
count   120975  120975.000000      120916       101400          50292   
unique      21            NaN         422         1204             17   
top         88            NaN  California  Napa Valley  Central Coast   
freq     16014            NaN       36104         4475          10975   
mean       NaN      35.363389         NaN          NaN            NaN   
std        NaN      41.022218         NaN          NaN            NaN   
min        NaN       4.000000         NaN          NaN            NaN   
25%        NaN      17.000000         NaN          NaN            NaN   
50%        NaN      25.000000         NaN          NaN            NaN   
75%        NaN      42.000000         NaN          NaN            NaN   
max        NaN    3300.000000         NaN          NaN            NaN   

       taster_name taster_twitter_handle  \
count        96479                 91559   
unique          19                    15   
top     Roger Voss            @vossroger   
freq         20172                 20172   
mean           NaN                   NaN   
std            NaN                   NaN   
min            NaN                   NaN   
25%            NaN                   NaN   
50%            NaN                   NaN   
75%            NaN                   NaN   
max            NaN                   NaN   

                                                    title     variety  \
count                                              120975      120974   
unique                                             110638         697   
top     Gloria Ferrer NV Sonoma Brut Sparkling (Sonoma...  Pinot Noir   
freq                                                   11       12787   
mean                                                  NaN         NaN   
std                                                   NaN         NaN   
min                                                   NaN         NaN   
25%                                                   NaN         NaN   
50%                                                   NaN         NaN   
75%                                                   NaN         NaN   
max                                                   NaN         NaN   

            winery  
count       120975  
unique       15855  
top     Testarossa  
freq           217  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN

## 1.2 Find the 10 varieties of wine which receives the highest number of  reviews

In [ ]:
# write your code here
# you may use functions such as value_counts()  
filtered_dat['variety'].value_counts().head(10)

Pinot Noir                  12787
Chardonnay                  11080
Cabernet Sauvignon           9386
Red Blend                    8476
Bordeaux-style Red Blend     5340
Riesling                     4972
Sauvignon Blanc              4783
Syrah                        4086
Rosé                         3262
Merlot                       3062
Name: variety, dtype: int64

## 1.3 Find varieties of wine having the average price less than 20, with the average pointsat least 90

In [ ]:
# write your code here
# you may use functions such as groupby() 
df=filtered_dat.groupby('variety').count()
df.sort_values(by='description', ascending=False).head(10).reset_index()['variety']
average_points = filtered_dat.groupby('variety').mean()
print(average_points)
filtered_dat['points']=filtered_dat['points'].astype(float)
filtered_dat.dtypes
top_average =average_points[(average_points['points']>=90) & (average_points['price']<20)]
print(top_average)


                            points      price
variety                                      
Abouriou                 87.666667  35.000000
Agiorgitiko              87.031746  23.571429
Aglianico                89.397959  38.887755
Aidani                   82.000000  27.000000
Airen                    81.666667   9.000000
Albana                   89.263158  23.473684
Albanello                86.000000  20.000000
Albariño                 87.725738  21.297468
Albarossa                88.000000  40.000000
Aleatico                 85.750000  39.750000
Alfrocheiro              88.687500  26.562500
Alicante                 87.400000  23.000000
Alicante Bouschet        89.458333  31.291667
Aligoté                  86.517241  19.655172
Alsace white blend       90.346939  29.571429
Altesse                  88.750000  17.875000
Alvarelhão               86.500000  21.000000
Alvarinho                88.525862  19.344828
Alvarinho-Chardonnay     86.750000  11.250000
Ansonica                 86.500000

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


## 1.4 Build statistic table 

In [ ]:
# write your code here
# # you may use functions such as groupby() and round(decimals=2)
statistic_table=filtered_dat.groupby('country').describe().round(2)
statistic_table

points                                                \
                          count   mean   std   min    25%   50%    75%    max   
country                                                                         
Argentina                3756.0  86.71  3.18  80.0  84.00  87.0  89.00   97.0   
Armenia                     2.0  87.50  0.71  87.0  87.25  87.5  87.75   88.0   
Australia                2294.0  88.60  3.00  80.0  87.00  89.0  91.00  100.0   
Austria                  2799.0  90.19  2.46  82.0  89.00  90.0  92.00   97.0   
Bosnia and Herzegovina      2.0  86.50  2.12  85.0  85.75  86.5  87.25   88.0   
Brazil                     47.0  84.66  2.45  80.0  83.00  85.0  86.00   89.0   
Bulgaria                  141.0  87.94  2.08  80.0  87.00  89.0  89.00   91.0   
Canada                    254.0  89.38  2.40  82.0  88.00  90.0  91.00   94.0   
Chile                    4416.0  86.50  2.70  80.0  85.00  86.0  88.00   95.0   
China                       1.0  89.00   NaN  89.0  89.00  89.0  89.00   89.0   
Croatia                    71.0  87.35  2.29  81.0  87.00  88.0  89.00   91.0   
Cyprus                     11.0  87.18  1.60  85.0  86.00  87.0  88.50   89.0   
Czech Republic             12.0  87.25  1.71  84.0  86.50  88.0  88.25   89.0   
England                    69.0  91.55  1.84  89.0  90.00  91.0  93.00   95.0   
France                  17776.0  88.73  3.01  80.0  87.00  88.0  91.00  100.0   
Georgia                    84.0  87.68  2.11  82.0  87.00  88.0  89.00   92.0   
Germany                  2120.0  89.84  2.47  81.0  88.00  90.0  91.00   98.0   
Greece                    461.0  87.29  2.19  82.0  86.00  87.0  89.00   93.0   
Hungary                   145.0  89.17  2.68  81.0  88.00  89.0  90.00   97.0   
India                       9.0  90.22  1.72  87.0  90.00  90.0  91.00   93.0   
Israel                    489.0  88.50  2.48  80.0  87.00  89.0  90.00   94.0   
Italy                   16914.0  88.62  2.66  80.0  87.00  88.0  90.00  100.0   
Lebanon                    35.0  87.69  2.78  82.0  86.50  88.0  90.00   91.0   
Luxembourg                  6.0  88.67  0.82  88.0  88.00  88.5  89.00   90.0   
Macedonia                  12.0  86.83  1.75  84.0  85.75  87.0  88.00   89.0   
Mexico                     70.0  85.26  2.72  80.0  83.00  85.0  87.00   92.0   
Moldova                    59.0  87.20  2.42  82.0  86.00  88.0  89.00   91.0   
Morocco                    28.0  88.57  1.69  85.0  88.00  88.5  89.25   93.0   
New Zealand              1378.0  88.31  2.44  80.0  87.00  88.0  90.00   95.0   
Peru                       16.0  83.56  1.86  80.0  82.00  84.0  85.00   86.0   
Portugal                 4875.0  88.32  3.02  80.0  86.00  88.0  91.00  100.0   
Romania                   120.0  86.40  1.72  82.0  85.00  86.0  87.25   92.0   
Serbia                     12.0  87.50  1.31  86.0  86.00  87.5  89.00   89.0   
Slovakia                    1.0  87.00   NaN  87.0  87.00  87.0  87.00   87.0   
Slovenia                   80.0  88.01  1.78  82.0  87.00  88.0  89.00   92.0   
South Africa             1293.0  87.83  2.35  82.0  86.00  88.0  90.00   95.0   
Spain                    6573.0  87.29  3.07  80.0  85.00  87.0  89.00   98.0   
Switzerland                 7.0  88.57  2.51  83.0  89.00  89.0  90.00   90.0   
Turkey                     90.0  88.09  1.98  84.0  87.00  88.0  90.00   92.0   
US                      54265.0  88.57  3.12  80.0  86.00  88.0  91.00  100.0   
Ukraine                    14.0  84.07  1.59  82.0  83.00  84.0  84.75   88.0   
Uruguay                   109.0  86.75  2.69  80.0  85.00  87.0  88.00   92.0   

                          price                                           \
                          count   mean    std   min    25%   50%     75%   
country                                                                    
Argentina                3756.0  24.51  23.43   4.0  12.00  17.0   25.00   
Armenia                     2.0  14.50   0.71  14.0  14.25  14.5   14.75   

In [ ]:
# save your table to 'statisticByState.csv'
statistic_table.to_csv('statisticByState.csv')

## 1.5 Recommendations


Based on the analysis, which country/countries would you recommend *HOTEL TULIP* to source wine from? Pleas state your reasons.

---

Your Answer:

- TO BE ADDED
    I recommend HOTEL TULIP to buy winery from  us and italy countries because they have the maximum number of reviews and their mean points are 88.57, 88.62 and their price are also affortable.


# Part 2. Text Analysis

## 2.1 extract high requency words in description

In [ ]:
import re
import nltk
from nltk.tokenize import RegexpTokenizer
from nltk.probability import *
from itertools import chain
#from tqdm import tqdm
import codecs

In [ ]:
with open('stopwords.txt') as f:
    stop_words = f.read().splitlines()
stop_words = set(stop_words)

In [ ]:
# write your code here
# define your tokenize

tokenizer = RegexpTokenizer(r"\w+(?:[-']\w+)?") 
highrequency_words = filtered_dat['description'].str.lower().str.cat(sep='\n')
unigram_tokens = tokenizer.tokenize(highrequency_words)
stopwords = set(stop_words)
final_tokens = [w for w in unigram_tokens if w not in stop_words]
final_tokens

['ripe',
 'fruity',
 'wine',
 'smooth',
 'structured',
 'firm',
 'tannins',
 'filled',
 'juicy',
 'red',
 'berry',
 'fruits',
 'freshened',
 'acidity',
 'drinkable',
 '2016',
 'tart',
 'snappy',
 'flavors',
 'lime',
 'flesh',
 'rind',
 'dominate',
 'green',
 'pineapple',
 'pokes',
 'crisp',
 'acidity',
 'underscoring',
 'flavors',
 'wine',
 'stainless-steel',
 'fermented',
 'pineapple',
 'rind',
 'lemon',
 'pith',
 'orange',
 'blossom',
 'start',
 'aromas',
 'palate',
 'bit',
 'opulent',
 'notes',
 'honey-drizzled',
 'guava',
 'mango',
 'giving',
 'slightly',
 'astringent',
 'semidry',
 'finish',
 'regular',
 'bottling',
 '2012',
 'rough',
 'tannic',
 'rustic',
 'earthy',
 'herbal',
 'characteristics',
 'nonetheless',
 'pleasantly',
 'unfussy',
 'country',
 'wine',
 'good',
 'companion',
 'hearty',
 'winter',
 'stew',
 'blackberry',
 'raspberry',
 'aromas',
 'show',
 'typical',
 'navarran',
 'whiff',
 'green',
 'herbs',
 'case',
 'horseradish',
 'mouth',
 'fairly',
 'full',
 'bodied',


In [ ]:
# find top common words with document frequencies > 5000
# you may use function FreqDist() and sort()
words = nltk.FreqDist(final_tokens)
words
frequent_words = set([k for k, v in words.items() if v > 5000])
HighFreq = sorted(frequent_words)
HighFreq

['acidity',
 'apple',
 'aromas',
 'balance',
 'balanced',
 'berry',
 'bit',
 'black',
 'blackberry',
 'blend',
 'bright',
 'cabernet',
 'character',
 'cherry',
 'chocolate',
 'citrus',
 'clean',
 'concentrated',
 'creamy',
 'crisp',
 'currant',
 'dark',
 'dried',
 'drink',
 'dry',
 'earthy',
 'fine',
 'finish',
 'firm',
 'flavor',
 'flavors',
 'fresh',
 'fruit',
 'fruits',
 'fruity',
 'full',
 'give',
 'good',
 'green',
 'herb',
 'herbal',
 'hint',
 'juicy',
 'lemon',
 'licorice',
 'light',
 'long',
 'made',
 'merlot',
 'mineral',
 'mouth',
 'nose',
 'note',
 'notes',
 'oak',
 'offers',
 'opens',
 'orange',
 'palate',
 'peach',
 'pear',
 'pepper',
 'pinot',
 'plum',
 'raspberry',
 'red',
 'rich',
 'ripe',
 'sauvignon',
 'savory',
 'shows',
 'smooth',
 'soft',
 'spice',
 'spicy',
 'structure',
 'style',
 'sweet',
 'syrah',
 'tannic',
 'tannins',
 'tart',
 'texture',
 'toast',
 'touch',
 'vanilla',
 'white',
 'wine',
 'wood',
 'years']

In [ ]:
# save your table to 'top_common_words.txt'  
with open("HighFreq.txt", 'a') as output:
  for frequent_words in HighFreq:
    print(output.write('%s\n' % frequent_words))


## 2.2 Find key words for describing Shiraz using TF-IDF

In [ ]:
# select 'description' from 'variety' eqaul to  'Shiraz' 

sh=filtered_dat[filtered_dat['variety']=='shiraz']['description']
sh


Series([], Name: description, dtype: object)

In [ ]:
# use TfidfVectorizer to calculate TF-IDF score
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf = TfidfVectorizer(analyzer='word', stop_words = 'english')
variable=tfidf.fit_transform(sh)
vec_idf=tfidf.idf_
tfids_result=dict(zip(tfidf.get_feature_names(),vec_idf))
tfids_result


In [ ]:
# find words with TF-IDF score >0.4 and sort them
tfifd=list((key,value.round(2)) for (key,value) in tfids_result.items() if value> 0.4)
tfifd.sort()
tfifd

In [ ]:
# save your table to 'key_Shiraz.txt'  
with open("Shirazkey.txt", 'a') as output:
  for frequent_words in HighFreq:
    print(output.write('%s\n' % frequent_words))